### PR TITLE
Support None from training_step in LRFinder

### PR DESCRIPTION
**Pull request by NikolasWolke**
Lightning modules may return None in the training_step, in order to skip the current batch when using automatic optimization. However, the learning rate finder failed in that case.

## What does this PR do?

This fix
* modifies the `_LRCallback.on_train_batch_end` method to support None from the training step by also handling an empty dict (since `_AutomaticOptimization.run` turns None `STEP_OUTPUT` into an empty dict)
* makes sure, that the list of losses and lrs have the same length as otherwise the selected learning rate would not be correct. If the lightning module returns None in the training step, `nan` is added to list of losses. because they are ignored when computing the suggestion.

Fixes https://github.com/Lightning-AI/lightning/issues/17992

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: tests/tests_pytorch/tuner/test_lr_finder.py [hunk 1]
```diff
@@ -503,3 +506,35 @@ def configure_optimizers(self):
 
     assert trainer.num_val_batches[0] == len(trainer.val_dataloaders)
     assert trainer.num_val_batches[0] != num_lr_tuner_training_steps
+
+
+def test_lr_finder_for_model_none_output(tmpdir):
```
Comment (awaelchli) [inline tests/tests_pytorch/tuner/test_lr_finder.py:511]:
```suggestion
def test_lr_finder_training_step_none_output(tmpdir):
```

**Review by awaelchli — APPROVED** [review]
Looks great, only minor comments!

File: tests/tests_pytorch/tuner/test_lr_finder.py [hunk 2]
```diff
@@ -503,3 +506,35 @@ def configure_optimizers(self):
...
+
+    seed_everything(1)
+    model = CustomBoringModel()
+
+    trainer = Trainer(default_root_dir=tmpdir)
+
+    tuner = Tuner(trainer)
+    # boring model runs into early stopping at step 95
+    # so disable to easier check we see the expected number of lrs and losses
+    lr_finder = tuner.lr_find(model=model, update_attr=True, num_training=100, early_stop_threshold=None)
```
[the remark this answers is not in the document]
Comment (NikolasWolke) [reply tests/tests_pytorch/tuner/test_lr_finder.py:534]: Yes, can reduce the number of steps. I can either go with 10 as suggested (and reduce `skip_begin` to say 3 when computing the suggestion) or use 20.

File: tests/tests_pytorch/tuner/test_lr_finder.py (hunk 1 above)
Comment (NikolasWolke) [reply tests/tests_pytorch/tuner/test_lr_finder.py]: Yes, that is better indeed. Changed it.

**Comment by NikolasWolke** [comment]
In the last build [pl-cpu (macOS-11, lightning, 3.10, 2.0)](https://github.com/Lightning-AI/lightning/actions/runs/5622495308/job/15235314528?pr=18129), [pl-cpu (ubuntu-20.04, lightning, 3.10, 2.0)](https://github.com/Lightning-AI/lightning/actions/runs/5622495308/job/15235314714?pr=18129) and [probot](https://github.com/Lightning-AI/lightning/actions/runs/5622495305/job/15235313094?pr=18129) failed all simply with `Error: `. Not sure what to make of that :thinking:

**Review by awaelchli — APPROVED** [review]
Thanks <3

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/tuner/lr_finder.py [hunk 3]
```diff
@@ -393,7 +393,11 @@ def on_train_batch_end(
         if (trainer.fit_loop.batch_idx + 1) % trainer.accumulate_grad_batches != 0:
             return
 
-        if outputs is None:
+        # _AutomaticOptimization.run turns None STEP_OUTPUT into an empty dict
+        if outputs is None or not outputs:
```
Comment (carmocca) [inline src/lightning/pytorch/tuner/lr_finder.py:397]: This would cover both cases
```suggestion
        if not outputs:
```

**Review by carmocca — APPROVED** [review]
